### PR TITLE
feat: ping/heartbeat (list devices)

### PR DESCRIPTION
**Pull request by xavierchanth**
**- What I did**

### Completed the following tasks from #184
- [x] sshnpd - internal schedule with ttl to create/refresh device key (json structure with devicename, metadata of installed version, and room for info)
- [x] sshnpd - monitor for ping notifications and respond with a device name notification
- [x] sshnp - grab key list for all clients
- [x] sshnp - ping notification and wait up to 5 seconds for response

In addition, adding some of my work on the python management scripts to help with testing this code.

Python scripts:

- Initial (experimental) SSHNP / SSH / SFTP Client written using paramiko
- A script which uses the aforementioned clients to update sshnpd installations through sshnp

**- How I did it**

**- How to verify it**

**- Description for the changelog**
feat: ping/heartbeat (list devices)

### CODE REVIEW

File: packages/sshnoports/lib/common/utils.dart [hunk 1]
```diff
@@ -45,13 +45,10 @@ Future<bool> fileExists(String file) async {
...
+const String asciiMatcher = r'[a-zA-Z0-9_]{0,15}';
+
 bool checkNonAscii(String test) {
-  var extra = test.replaceAll(RegExp(r'[a-zA-Z0-9_]*'), '');
-  if ((extra != '') || (test.length > 15)) {
-    return true;
-  } else {
-    return false;
-  }
+  return RegExp(asciiMatcher).allMatches(test).first.group(0) != test;
```
Comment (xavierchanth) [inline packages/sshnoports/lib/common/utils.dart:51]: The first match will always be the whole thing (if it matches), so we just compare the first match to test, if there are no matches, the left always resolves to empty string.

**Comment by gkc** [comment]
@XavierChanth Looks good! Will you be adding test(s) to the e2e test pack? Also while I was doing my code walkthrough I noticed something we missed during the refactorings and I pushed [a commit](url) to address that

I will not approve and merge yet as the PR is marked as a draft.

PS Some checks failed with
```
sshnpd  | SEVERE|2023-07-26 11:05:10.923584|OnboardingCli|Caught exception: Exception: Failed connecting to root server url root.atsign.org on port 64
```
I will re-run the checks

**Comment by xavierchanth** [comment]
CleanShot 2023-07-29 at 08 20 53@2x

playpen is my running instance, 
dead is an instance I spun up and then killed to see if the atKey was put in place correctly

Then I switched to interiorwolverine for the device atSign (which is not running any sshnpds) and it shows no devices, as well as a prompt hinting that they should try updating sshnpd

**Comment by xavierchanth** [comment]
Made the output a little bit nicer: 
CleanShot 2023-07-29 at 08 25 47@2x

**Comment by xavierchanth** [comment]
I did not complete the task that ensures device names are unique, Want to make sure we do that one right. Initially we thought of binding a device name to a client ID, but we also have the option of preventing subsequent sshnpd startups if a device name is already in use. i.e. check for an atKey and if there is an atKey try to do a ping (probably 2-3 times to be safe) then startup sshnpd

**Review by gkc — APPROVED** [review]

**Comment by xavierchanth** [comment]
@gkc did you confirm this to be working before merging? I still had some testing to do

**Comment by gkc** [comment]
I did not; I only reviewed the code, it looked complete as was not draft PR and tasks had been ticked off 

But now that you mention it I do now remember that you had said something about it being tricky to test

Let's just create another PR for follow-up testing & test automation

**Comment by xavierchanth** [comment]
> I did not; I only reviewed the code, it looked complete as was not draft PR and tasks had been ticked off
> 
> But now that you mention it I do now remember that you had said something about it being tricky to test
> 
> Let's just create another PR for follow-up testing & test automation

My mistake, maybe I did complete testing... since I marked it as ready for review... I honestly don't remember how far along I got with testing. It never hurts to test twice though.

**Pull request merged**